product_variant_change_attribute_value/wizards/product_variant_attribute_value_wizard.py

```diff
@@ -0,0 +1,256 @@
+            else:
+                raise
+
+    def _cleanup_attribute_values(self, product, pavs_to_clean):
+        TplAttrValue = self.env["product.template.attribute.value"]
+        template = product.product_tmpl_id
+        for attr, pavs in pavs_to_clean.items():
+            tpl_attr_line = template.attribute_line_ids.filtered(
+                lambda l: l.attribute_id == attr
+            )
+            # Ensure that product variant combinations are not created
+            # during cleanup.
+            tpl_attr_line = tpl_attr_line.with_context(
+                update_product_template_attribute_values=False
+            )
+            error_msg = self._unique_err_msg(product, tpl_attr_line, pavs)
+            if not (tpl_attr_line.value_ids - pavs):
+                # no value left
+                def _make_inactive():
+                    tpl_attr_line.active = False
+
+                self._handle_unique_violation(_make_inactive, error_msg)
+            tpl_attr_line.write({"value_ids": [(3, pav.id) for pav in pavs]})
+            tpl_attr_values = TplAttrValue.search(
+                [
+                    ("attribute_line_id", "=", tpl_attr_line.id),
+                    ("product_attribute_value_id", "in", pavs.ids),
+                ]
+            )
+            if tpl_attr_values:
+                self._handle_unique_violation(tpl_attr_values.unlink, error_msg)
+
+    def _remove_duplicate_product(self, product):
+        product_removed = False
+        for check_product in self.product_ids - product:
+            variant = set(check_product.product_template_attribute_value_ids.ids)
+            product_temp = set(product.product_template_attribute_value_ids.ids)
+            if variant.issubset(product_temp):
+                product_used = self._is_product_exist_models(product)
+                check_product_used = self._is_product_exist_models(check_product)
+                if not product_used:
+                    product_removed = product.unlink()
+                elif product_used and not check_product_used:
+                    check_product.unlink()
+                else:
+
+                    raise UserError(
+                        _(
+                            "The product %s associated \
+                                with Sale Orders/Invoices, impossoble to remove"
+                            % (product.display_name)
+                        )
+                    )
+        return product_removed
+
+    def _is_product_exist_models(self, product):
+        used = False
+        models = self.env["ir.model"].search(
+            [("model", "in", ["account.move.line", "sale.order.line"])]
+        )
+        for model in models:
+            used = self._is_product_exist_in_record(model.model, product)
+        return used
+
+    def _is_product_exist_in_record(self, model, product):
+        return self.env[model].search([("product_id", "=", product.id)], limit=1)
```

Comment:
May I know the benefit of using additional method instead of below?
```suggestion
        domain = [("product_id", "=", product.id)]
        for model in models:
            if self.env[model.model].search(domain, limit=1):
                return True
        return False
```

Reply:
Branch updated, thanks